README.md
# ngCroppie - Angular Croppie Tool

[![Build Status][travis-image]][travis-url]
[![Bower version][bower-image]][bower-url]
[![NPM version][npm-image]][npm-url]
[![Dependency Status][depstat-image]][depstat-url]
[![NPM Montly Downloads][npm-downloads-image]][npm-downloads-url]
[![MIT License][license-image]][license-url]

> Angular Croppie Tool is an image cropping and rotating module for AngularJS based on [Croppie.js](https://github.com/Foliotek/Croppie/)<br>
Pure Javascript implementation; Also responsive!

## Dependencies
- [Angular](https://github.com/angular/angular.js)
- [Croppie.js](https://github.com/Foliotek/Croppie/)

## Installation

```sh
# Using bower:
$ bower install ngCroppie

# Using npm:
$ npm install ng-croppie
```

## Basic usage
- Include `ngCroppie` module into your project;
- No dependencies are required in the controller;
- Add `<ng-croppie>` tag with the following arguments:
  - `src` **Image** or **Blob**: path to Image File or Base64;
  - `ng-model` **Base64**: the image output;

#### [Demo](http://orif-jr.github.io/ng-croppie/#demo)

## Example
**JS**
```js
function ExampleCtrl() {
  var vm = this;

  vm.inputImage = null;
  vm.outputImage = null;

  vm.onUpdate = function(data) {
    //console.log(data);
  };
}

angular
  .module('exampleModule', ['ngCroppie'])
  .controller('ExampleCtrl', [ExampleCtrl]);
```

**HTML**
```html
<ng-croppie   src="inputImage"
              ng-model='outputImage'
              update='onUpdate'
              boundry="{w: 400, h: 400}"
              viewport="{w: 300, h: 300}"
              orientation="true"
              rotation="90"
              type="circle">
</ng-croppie>



<!-- Preview -->
<img ng-src="{{outputImage}}" />
```




## Parameters
- **Image** or **Blob** `data`: path to image file or base64;
- **Base64** `ng-model`: the image output, returns are a base64;
- **Function** `update`: place a functon in the controller to run whenever changes are made to the image;
- **Object** `boundry { w: __, h: __ }`: (optional) the size of the container that will host the croppie tool - defaults to 400x400;
- **Object** `viewport { w: __, h: __ }`: (optional) the output size, must be smaller than the boundry or it will equal it - defaults to 300x300;
- **Boolean** `zoom`: (optional) enable zoom slider - defaults to true;
- **Boolean** `mousezoom`: enable mouse scroll bar to zoom in/out. Works with `zoom` - defaults to true;
- **Boolean** `zoomslider`: enable zoom slider only (scrolling and pinching zoom still possible if set to false), works with `zoom` - defaults to true;
- **Boolean** `exif`: (optional) enable exif orientation compatability - defaults to undefined;
- **Boolean** `orientation`: (optional) enable custom orientation support when binding images - defaults to false.
- **Integer** `rotation`: rotate the image by a specified degree amount, works with `orientation` - valid values: 90, 180, 270, -90, -180, -270;
- **String** `type`: (optional) can either be 'circle' or 'square' - defaults to 'square';
- **Boolean** `mobile`: (optional) enable mobile view - defaults to viewport: 250x250 and boundry: 300x300;

## Version
1.2.3

### Updates
- update Croppie.js v2.6.2
- add AngularJS v1.6 dependency
- add Unit Tests

### Contributors

- orif-jr
- stasinua
- htao00
- alanheppenstall

### License
MIT © [ngCroppie](https://github.com/allenRoyston/ngCroppie#readme)

[travis-url]: https://travis-ci.org/orif-jr/ngCroppie
[travis-image]: https://img.shields.io/travis/orif-jr/ngCroppie.svg

[bower-url]: https://github.com/allenRoyston/ngCroppie#readme
[bower-image]: https://img.shields.io/bower/v/ngCroppie.svg

[npm-url]: https://npmjs.org/package/ng-croppie
[npm-image]: https://img.shields.io/npm/v/ng-croppie.svg

<!--
from david-dm
[depstat-url]: https://david-dm.org/allenRoyston/ngCroppie
[depstat-image]: https://david-dm.org/allenRoyston/ngCroppie.svg
-->
[depstat-url]: https://gemnasium.com/github.com/allenRoyston/ngCroppie
[depstat-image]: https://gemnasium.com/badges/github.com/allenRoyston/ngCroppie.svg
<!--
[![DevDependencies Status][devdepstat-image]][devdepstat-url]
[devdepstat-url]: https://david-dm.org/allenRoyston/ngCroppie?type=dev
[devdepstat-image]: https://david-dm.org/allenRoyston/ngCroppie/dev-status.svg
-->
[npm-downloads-url]: https://www.npmjs.com/package/ng-croppie
[npm-downloads-image]: https://img.shields.io/npm/dm/ng-croppie.svg

[license-image]: https://img.shields.io/badge/license-MIT-blue.svg
[license-url]: LICENSE


CHANGELOG.md
<a name="1.2.3"></a>
## [1.2.3](https://github.com/allenRoyston/ngCroppie/compare/1.2.2...1.2.3) (2018-12-04)

### Features
* update Croppie.js v2.6.2 ([a1b2c3d](https://github.com/allenRoyston/ngCroppie/commit/a1b2c3d))
* add AngularJS v1.6 dependency ([a1b2c3d](https://github.com/allenRoyston/ngCroppie/commit/a1b2c3d))
* add unit tests ([a1b2c3d](https://github.com/allenRoyston/ngCroppie/commit/a1b2c3d))


<a name="1.2.2"></a>
## [1.2.2](https://github.com/allenRoyston/ngCroppie/compare/1.2.1...1.2.2) (2018-01-11)

### Features
* add changelog file ([b1c84e6](https://github.com/allenRoyston/ngCroppie/commit/b1c84e6))

### Bug Fixes
* fix source interval time for firefox ([75d831f](https://github.com/allenRoyston/ngCroppie/commit/75d831f))
* fix travis build badge ([2aa5c3e](https://github.com/allenRoyston/ngCroppie/commit/2aa5c3e))


<a name="1.2.1"></a>
## [1.2.1](https://github.com/allenRoyston/ngCroppie/compare/1.2.0...1.2.1) (2017-12-19)

### Features
* add new feature **enforceBoundary** ([5785524](https://github.com/allenRoyston/ngCroppie/commit/5785524))
* update dev dependencies ([0e534da](https://github.com/allenRoyston/ngCroppie/commit/0e534da))


<a name="1.2.0"></a>
## [1.2.0](https://github.com/allenRoyston/ngCroppie/compare/1.0.3...1.2.0) (2017-12-17)

### Features
* library upgrade @Croppie.js v2.5.1 ([1f79735](https://github.com/allenRoyston/ngCroppie/commit/1f79735))

### Bug Fixes
* fix scroll wheel for firefox browser ([855b571](https://github.com/allenRoyston/ngCroppie/commit/855b571))


<a name="1.0.3"></a>
## [1.0.3](https://github.com/allenRoyston/ngCroppie/compare/1.0.2...1.0.3) (2017-07-24)

### Features
* library upgrade Croppie.js v2.5.0 ([f58fefb](https://github.com/allenRoyston/ngCroppie/commit/f58fefb))
* add new feature **mobile** ([280993e](https://github.com/allenRoyston/ngCroppie/commit/280993e))
* update dev dependencies ([5cb030b](https://github.com/allenRoyston/ngCroppie/commit/5cb030b))

### Bug Fixes
* fix source watcher ([4b25e6a](https://github.com/allenRoyston/ngCroppie/commit/4b25e6a))


<a name="1.0.2"></a>
## [1.0.2](https://github.com/allenRoyston/ngCroppie/compare/1.0.1...1.0.2) (2017-05-15)

### Bug Fixes
* fix watchers for source and rotation ([fa982f7](https://github.com/allenRoyston/ngCroppie/commit/fa982f7)),  closes [#17](https://github.com/allenRoyston/ngCroppie/issues/17)


<a name="1.0.1"></a>
## [1.0.1](https://github.com/allenRoyston/ngCroppie/compare/1.0.0...1.0.1) (2017-04-06)

### Features
* library upgrade Croppie.js v2.4.1 ([a708ac0](https://github.com/allenRoyston/ngCroppie/commit/a708ac0))
* add new feature **zoomslider** ([a708ac0](https://github.com/allenRoyston/ngCroppie/commit/a708ac0))
* add new features **zoomer**, **orientation** and **exif** ([550f036](https://github.com/allenRoyston/ngCroppie/commit/550f036))

### Bug Fixes
* code modularizing & refactoring ([1d3d1f2](https://github.com/allenRoyston/ngCroppie/commit/1d3d1f2)), closes [#6](https://github.com/allenRoyston/ngCroppie/issues/6)
* zoom, mousezoom and rotate ([550f036](https://github.com/allenRoyston/ngCroppie/commit/550f036))
* default zoom ([6a34e83](https://github.com/allenRoyston/ngCroppie/commit/550f036))


<a name="1.0.0"></a>
## [1.0.0](https://github.com/allenRoyston/ngCroppie/compare/ab9e64c...1.0.0) (2017-03-31)

### Notes
* initial release, closes [#4](https://github.com/allenRoyston/ngCroppie/issues/4), [#12](https://github.com/allenRoyston/ngCroppie/issues/12), [#15](https://github.com/allenRoyston/ngCroppie/issues/15) and [#26](https://github.com/allenRoyston/ngCroppie/issues/26)


gulpfile.js
var gulp = require('gulp');
var concat = require('gulp-concat');
var pump = require('pump');
var mainBowerFiles = require('gulp-main-bower-files');
var minify = require('gulp-clean-css');
var rename = require('gulp-rename');
var uglify = require('gulp-uglify');


gulp.task('css', function(cb) {
    pump([
        gulp.src('./bower.json'),
        mainBowerFiles('**/*.css'),
        concat('src/ng-croppie.css'),
        gulp.dest('./')
    ], cb);
});

gulp.task('css-minify', function(cb) {
    pump([
        gulp.src('./bower.json'),
        mainBowerFiles('**/*.css'),
        concat('src/ng-croppie.css'),
        rename({suffix: '.min'}),
        minify({compatibility: 'ie11', keepBreaks: true}),
        gulp.dest('./')
    ], cb);
});

gulp.task('js', function(cb) {
    pump([
        gulp.src('src/ng-croppie.js'),
        rename({suffix: '.min'}),
        uglify(),
        gulp.dest('./src')
    ], cb);
});

gulp.task('default', ['css', 'css-minify', 'js']);

src/ng-croppie.js
/*!
 *	 Angular Croppie Tool (ngCroppie)
 *	 An awesome image cropping and rotating module for AngularJS.
 *
 *	 Credits: https://github.com/allenRoyston/ngCroppie/graphs/contributors
 *	 Inspired by Croppie.js https://github.com/Foliotek/Croppie
 *
 *	 Version: 1.2.3
 * 	 License: MIT
 */
(function () {
    'use strict';

    var module = angular.module('ngCroppie', []);

    module.directive('ngCroppie', ['$timeout', function ($timeout) {
        return {
            restrict: 'AE',
            scope: {
                src: '=',
                rotation: '=',
                viewport: '=',
                boundry: '=',
                type: '@',
                zoom: '@',
                mousezoom: '@',
                zoomslider: '@',
                exif: '@',
                enforceBoundary: '@',
                orientation: '@',
                update: '=',
                ngModel: '=',
                mobile: '@'
            },
            link: function (scope, elem, attr) {
                // defaults
                if (typeof scope.viewport === 'undefined') {
                    scope.viewport = { w: null, h: null };
                }
                if (typeof scope.boundry === 'undefined') {
                    scope.boundry = { w: null, h: null };
                }

                // catches
                if (scope.mobile === 'true') {
                    scope.viewport.w = 250; scope.viewport.h = 250;
                    scope.boundry.w = 300; scope.boundry.h = 300;
                } else {
                    scope.viewport.w = (typeof scope.viewport.w !== 'undefined') ? scope.viewport.w : 300;
                    scope.viewport.h = (typeof scope.viewport.h !== 'undefined') ? scope.viewport.h : 300;
                    scope.boundry.w = (typeof scope.boundry.w !== 'undefined') ? scope.boundry.w : 400;
                    scope.boundry.h = (typeof scope.boundry.h !== 'undefined') ? scope.boundry.h : 400;
                }

                // viewport cannot be larger than the boundaries
                if (scope.viewport.w > scope.boundry.w) {
                    scope.viewport.w = scope.boundry.w
                }
                if (scope.viewport.h > scope.boundry.h) {
                    scope.viewport.h = scope.boundry.h
                }

                // convert string to Boolean
                var zoom = (scope.zoom === 'true' || typeof scope.zoom === 'undefined'),
                    mouseZoom = (scope.mousezoom === 'true' || typeof scope.mousezoom === 'undefined'),
                    zoomSlider = (scope.zoomslider === 'true' || typeof scope.zoomslider === 'undefined');

                // define options
                var options = {
                    viewport: {
                        width: scope.viewport.w,
                        height: scope.viewport.h,
                        type: scope.type || 'square'
                    },
                    boundary: {
                        width: scope.boundry.w,
                        height: scope.boundry.h
                    },
                    enableZoom: zoom,
                    mouseWheelZoom: mouseZoom,
                    showZoomer: zoomSlider,
                    enableExif: scope.exif,
                    enforceBoundary: scope.enforceBoundary,
                    enableOrientation: scope.orientation
                };

                if (typeof scope.update !== 'undefined') {
                    options.update = scope.update;
                }

                // create new croppie and settime for updates
                var c = new Croppie(elem[0], options);
                // get Croppie elements for further calculations
                var croppieBody = angular.element(elem[0])[0];
                var croppieCanvas = angular.element(elem[0].getElementsByClassName('cr-boundary'))[0];

                var intervalID;

                // TODO: needs for investigation
                var croppieCanvasRectangle = croppieCanvas.getBoundingClientRect();

                // initialize interval only if action registered within ngCroppie container
                croppieBody.addEventListener('mousedown', function() {
                    intervalID = window.setInterval(function() {
                        c.result('canvas').then(function(img) {
                            scope.$apply(function() {
                                scope.ngModel = img;
                            });
                        });
                    }, 100);
                }, false);

                // check mouseZoom property to avoid needless event listener initialization
                // separated "wheel" event listener to prevent conflict with Croppie default "wheel" event listener
                if (mouseZoom) {
                    // IE9, Chrome, Opera, Safari
                    croppieBody.addEventListener('mousewheel', function(evt) {
                        evt.preventDefault();
                        c.result('canvas').then(function(img) {
                            scope.$apply(function() {
                                scope.ngModel = img;
                            });
                        });
                    }, false);

                    // Firefox
                    croppieBody.addEventListener('DOMMouseScroll', function(evt) {
                        evt.preventDefault();
                        c.result('canvas').then(function(img) {
                            scope.$apply(function() {
                                scope.ngModel = img;
                            });
                        });

                    }, false);
                }

                // destroy all created intervals
                croppieBody.addEventListener('mouseup', function() {
                    clearInterval(intervalID);
                }, false);
                croppieBody.addEventListener('mouseleave', function() {
                    clearInterval(intervalID);
                }, false);
                croppieBody.addEventListener('mouseout', function() {
                    clearInterval(intervalID);
                }, false);

                scope.$on('$destroy', function (event) {
                    clearInterval(intervalID);
                });

                // image rotation
                scope.$watch('rotation', function(newValue, oldValue) {
                    if (scope.orientation === 'false' || typeof scope.orientation === 'undefined') {
                        throw 'ngCroppie: Cannot rotate without \'orientation\' option';
                    } else {
                        c.rotate(newValue - oldValue);
                        c.result('canvas').then(function(img) {
                            scope.$apply(function () {
                                scope.ngModel = img;
                            });
                        });
                    }
                });

                // respond to changes in src
                scope.$watch('src', function(newValue, oldValue) {
                    if (typeof scope.src === 'undefined') {
                        throw 'ngCroppie: Image source undefined!'
                    } else {
                        c.bind(scope.src);
                        window.setInterval(function() {  // force delay for the ng-file-upload
                            c.result('canvas').then(function(img) {
                                scope.$apply(function () {
                                    scope.ngModel = img;
                                });
                            });
                        }, 0);
                    }
                });
            }
        };
    }]);

}());
